### PR TITLE
deal with dedupe use case 2: log duplicate uploads for same healthcod…

### PR DESCRIPTION
**Pull request by zaclinsage**
…e, createdOn and schemaId

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java [hunk 1]
```diff
@@ -91,4 +98,15 @@ public HealthDataRecord getRecordById(@Nonnull String id) {
...
+    /** {@inheritDoc} */
+    @Override
+    public List<HealthDataRecord> getRecordsByHealthCodeCreatedOnSchemaId(@Nonnull String healthCode, @Nonnull Long createdOn, @Nonnull String schemaId) {
+        List<HealthDataRecord> recordsByHealthCode = healthCodeIndex.queryKeys(HealthDataRecord.class, "healthCode",
+                healthCode, null);
+
+        return recordsByHealthCode.stream()
+                .filter(record -> record.getCreatedOn().equals(createdOn) && record.getSchemaId().equals(schemaId))
+                .collect(Collectors.toList());
+    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java:149]: I think this needs to be a bit broader than that. Some duplicate uploads might not come in at the exact same time, but rather spaced a few seconds or a few minutes apart. You'll want a start and end time and use a range query.

Also, this query is very poor performing. Your query gets all the records for a particular healthcode, which could be hundreds if not thousands of records.

You also call queryKeys(), which only gets they keys (healthcode, recordId), which means schemaId and createdOn will be blank. You'll want to just call query().

Lastly, you'll want a unit test that calls through to the real DynamoDB, so you can make sure your code works against the actual DDB, and not a mock DDB that works the way you think it does (but might not actually). See for example DynamoUploadSchemaDaoDdbTest vs DynamoUploadSchemaDaoMockTest.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java [hunk 2]
```diff
@@ -46,14 +46,12 @@
     private ExporterStatus synapseExporterStatus;
 
     /** {@inheritDoc} */
-    @JsonSerialize(using = DateTimeToLongSerializer.class)
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java:49]: Why is are the JSON serializers and deserializers being removed from this class (here and in the other 3 cases in this file)? This will break the JSON format of the health data record.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 3]
```diff
@@ -22,6 +26,15 @@
 
     private List<UploadValidationHandler> handlerList;
     private UploadDao uploadDao;
+    private HealthDataService healthDataService;
+
+    public void setHealthDataService(HealthDataService healthDataService) {
+        this.healthDataService = healthDataService;
+    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java:33]: Make this final. Otherwise, subclasses my accidentally override this and lead to mysterious NullPointerExceptions.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 4]
```diff
@@ -111,6 +124,30 @@ public void run() {
         }
 
         // TODO: if validation fails, wipe the files from S3
+
+        // dedupe logic over here:
+        dedupeHelper(context.getRecordId());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java]: Play doesn't log exceptions from async threads. You need a try-catch here.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 5]
```diff
@@ -111,6 +124,30 @@ public void run() {
...
+    }
+
+    /**
+     * helper method: query healthdatarecord by healthcode, schema and createdOn from the upload just finished above,
+     * if there are more than 1 such records in ddb, there are duplicates. log that information but not stop or delete anything right now
+     * @param healthRecordId
+     */
+    private void dedupeHelper(String healthRecordId) {
+        // get necessary information for query first
+        HealthDataRecord record = healthDataService.getRecordById(healthRecordId);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java:152]: There might not be a record for this upload. (For example, if the upload failed validation.) What does your code do in this case?

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 6]
```diff
@@ -111,6 +124,30 @@ public void run() {
...
+        Long createdOn = record.getCreatedOn();
+        String healthCode = record.getHealthCode();
+        String schemaId = record.getSchemaId();
+
+        List<HealthDataRecord> retList = healthDataService.getRecordsByHealthcodeCreatedOnSchemaId(healthCode, createdOn, schemaId);
+
+        if (retList.size() > 1) {
+            logger.warn(String.format("Duplicate health data records for record id: %s, health code: %s, created on:  %s, schema id: %s, duplicate size: %s",
+                    healthRecordId, healthCode, createdOn, schemaId, retList.size()));
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java:184]: It's not enough to simply log that a duplicate exists. If we want to analyze, we need to actually identify the duplicates.

Also, this should be info, not warning. Warnings mean there's an immediately actionable if there are lots of them. This doesn't qualify.

Lastly, you'll want a unit test to make sure this works. It might be worth factoring the warning message to a "handleDuplicate()" method that you can spy and verify that it was called. You'll also want to test the cases where there is no record for this upload, where the record is not a duplicate, where there is a duplicate, and where there's more than one duplicate.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTaskFactory.java [hunk 7]
```diff
@@ -33,6 +35,11 @@ public void setUploadDao(UploadDao uploadDao) {
         this.uploadDao = uploadDao;
     }
 
+    @Autowired
+    public void setHealthDataService(HealthDataService healthDataService) {
+        this.healthDataService = healthDataService;
+    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTaskFactory.java:41]: Similarly, make this final.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Also, how did you test this? Unit tests? Integration tests? Manual tests? Testing a few cases manually and verifying the log message(s) would definitely help make sure your code works as intended.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java [hunk 8]
```diff
@@ -16,9 +26,12 @@
 /** DynamoDB implementation of {@link org.sagebionetworks.bridge.dao.HealthDataDao}. */
 @Component
 public class DynamoHealthDataDao implements HealthDataDao {
+    private static final long CREATED_ON_OFFSET_MILIS = 100;
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java]: I think this is too narrow. We can have dupes that are submitted seconds or even minutes apart. I'd suggest setting this to an hour or so.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java [hunk 9]
```diff
@@ -44,6 +57,16 @@ public void setUploadDateIndex(DynamoIndexHelper uploadDateIndex) {
         this.uploadDateIndex = uploadDateIndex;
     }
 
+    /**
+     * DynamoDB Index reference for the healthCodeCreatedOnIndex index. This is needed because the DynamoDB mapper does allow queries
+     * using global secondary indices. This is configured by Spring
+     */
+    @Resource(name = "healthDataHealthCodeCreatedOnIndex")
+    public void setHealthCodeCreatedOnIndex(DynamoIndexHelper healthCodeCreatedOnIndex) {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java]: Make this final.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java [hunk 10]
```diff
@@ -91,4 +114,37 @@ public HealthDataRecord getRecordById(@Nonnull String id) {
...
+        Condition rangeKeyCondition = new Condition()
+                .withComparisonOperator(ComparisonOperator.BETWEEN)
+                .withAttributeValueList(new AttributeValue().withN(String.valueOf((createdOn - CREATED_ON_OFFSET_MILIS)))
+                        , new AttributeValue().withN(String.valueOf((createdOn + CREATED_ON_OFFSET_MILIS))));
+
+        DynamoHealthDataRecord queryRecord = new DynamoHealthDataRecord();
+        queryRecord.setHealthCode(healthCode);
+
+        DynamoDBQueryExpression<DynamoHealthDataRecord> expression = new DynamoDBQueryExpression<DynamoHealthDataRecord>()
+                .withConsistentRead(false)
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java:128]: Why are we turning off consistent reads? If two duplicates come in at near the same time, we want to be be able to catch it.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 11]
```diff
@@ -1,30 +1,40 @@
...
+import java.util.List;
+
 @ContextConfiguration("classpath:test-context.xml")
 @RunWith(SpringJUnit4ClassRunner.class)
 public class DynamoHealthDataDaoDdbTest {
     private static final long CREATED_ON = 1424136378727L;
-    private static final String HEALTH_CODE = "health-code";
-    private static final String RECORD_ID = "record-id";
+    private static final String HEALTH_CODE = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
+    private static final String RECORD_ID = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: This almost certainly doesn't do what you want it to do. This creates the random name once during the initialization of the JVM and uses the same ID for all tests. This means there's a chance that tests might interfere with each other.

You're better off making this non-static and adding the initialization to an @Before method.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 12]
```diff
@@ -111,6 +122,61 @@ public void run() {
         }
 
         // TODO: if validation fails, wipe the files from S3
+
+        // dedupe logic over here:
+        try {
+            dedupeHelper(context.getRecordId());
+        } catch (ClassNotFoundException e) {
+            logClassNotFoundException(e);
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java:133]: I think there are more exceptions than just ClassNotFoundException. In theory, you need to be able to handle any and all exception types, since the ExceptionInterceptor doesn't work on async threads.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 13]
```diff
@@ -111,6 +122,61 @@ public void run() {
...
+     * @param healthRecordId
+     */
+    private void dedupeHelper(String healthRecordId) throws ClassNotFoundException {
+        // get necessary information for query first
+        HealthDataRecord record = healthDataService.getRecordById(healthRecordId);
+
+        // if there is no record in ddb,
+        if (record == null) {
+            throw new ClassNotFoundException("Record not found in ddb");
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java:157]: It's a perfectly normal scenario for an upload to not have a record, such as when validation fails. We shouldn't throw an exception here.

Also, healthDataService.getRecordById() will throw a Bad Request Exception if there is no record ID, so we need to handle that case as well.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 14]
```diff
@@ -111,6 +122,61 @@ public void run() {
...
+        List<HealthDataRecord> retList = healthDataService.getRecordsByHealthcodeCreatedOnSchemaId(healthCode, createdOn, schemaId);
+
+        if (retList.size() > 1) {
+            logger.info(String.format("Duplicate health data records for record id: %s, health code: %s, created on:  %s, schema id: %s, duplicate size: %s",
+                    healthRecordId, healthCode, createdOn, schemaId, retList.size()));
+
+            logDuplicateUploadRecords(retList);
+        } else if (retList.size() == 0) {
+            // should not be reached since getRecordById() already test if there is a record exists
+            throw new ClassNotFoundException("Record not found in ddb");
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java]: Do we need this else if check here? Is ClassNotFoundException the right exception to throw here? Is there any particular reason we're throwing an exception instead of logging?

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 15]
```diff
@@ -111,6 +122,61 @@ public void run() {
...
+
+        Long createdOn = record.getCreatedOn();
+        String healthCode = record.getHealthCode();
+        String schemaId = record.getSchemaId();
+
+        List<HealthDataRecord> retList = healthDataService.getRecordsByHealthcodeCreatedOnSchemaId(healthCode, createdOn, schemaId);
+
+        if (retList.size() > 1) {
+            logger.info(String.format("Duplicate health data records for record id: %s, health code: %s, created on:  %s, schema id: %s, duplicate size: %s",
+                    healthRecordId, healthCode, createdOn, schemaId, retList.size()));
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java]: Don't log the health code. This is a potential security/privacy risk.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 16]
```diff
@@ -111,6 +122,61 @@ public void run() {
...
+        // else there is only one such record exists in ddb, means no duplicate -- do nothing
+    }
+
+    void logDuplicateUploadRecords(List<HealthDataRecord> dupeRecords) {
+        for (HealthDataRecord dupeRecord: dupeRecords) {
+            logger.info("Duplicate HealthDataRecord: " + "record id: " + dupeRecord.getId()
+                    + ", createdOn: " + dupeRecord.getCreatedOn().toString()
+                    + ", healthCode: " + dupeRecord.getHealthCode() + ", schemaId: " + dupeRecord.getSchemaId()
+                    + ", studyId: " + dupeRecord.getStudyId() + ", uploadDate: " + dupeRecord.getUploadDate().toString()
+                    + ", uploadId: " + dupeRecord.getUploadId());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java]: Similarly, don't log the healthCode here either.

Also, the original record ID and duplicate record ID should be on the same line, to make log analysis easier.

Also, we need to make sure that we aren't logging records as duplicates of themselves.

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 17]
```diff
@@ -3,37 +3,129 @@
...
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
-import static org.mockito.Mockito.doThrow;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.notNull;
-import static org.mockito.Mockito.spy;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.*;
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java]: Code convention is to list out all imports we're using instead of using *

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 18]
```diff
@@ -3,37 +3,129 @@
...
 import org.sagebionetworks.bridge.dynamodb.DynamoUpload2;
+import org.sagebionetworks.bridge.json.BridgeObjectMapper;
+import org.sagebionetworks.bridge.models.healthdata.HealthDataRecord;
+import org.sagebionetworks.bridge.models.healthdata.HealthDataRecordBuilder;
 import org.sagebionetworks.bridge.models.upload.UploadStatus;
+import org.sagebionetworks.bridge.services.HealthDataService;
 
 public class UploadValidationTaskTest {
+    public UploadValidationTaskTest() throws IOException {
+    }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java]: Tests shouldn't have constructors.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java [hunk 19]
```diff
@@ -91,4 +114,37 @@ public HealthDataRecord getRecordById(@Nonnull String id) {
...
+                // This should never happen, but just in case.
+                throw new BridgeServiceException(String.format(
+                        "DynamoDB returned objects of type %s instead of %s",
+                        oneResult.getClass().getName(), DynamoHealthDataRecord.class.getName()));
+            }
+
+            finalRecords.add((HealthDataRecord) oneResult);
+        }
+
+        return finalRecords;
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java]: Where do we filter on schema ID?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 20]
```diff
@@ -86,4 +99,62 @@ public void ddbRoundtrip() throws Exception {
             mapper.delete(savedRecord);
         }
     }
+
+    @Test
+    public void getRecordsByHealthCodeCreatedOnSchemaId() throws IOException {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: We should also add tests for:
* different health code
* createdOn out of range
* different schema ID

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 21]
```diff
@@ -150,6 +244,130 @@ public void writeValidationStatusException() {
...
+
+        // execute
+        task.run();
+
+        // verify log helper was NOT called
+        verify(task, times(0)).logDuplicateUploadRecords(any());
+    }
+
+    @Test
+    public void dedupeNullRecord() {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java:341]: Should we also have a test case where upload.getRecordId() is null?

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 22]
```diff
@@ -150,6 +244,130 @@ public void writeValidationStatusException() {
...
+
+        // set up validation task
+        UploadValidationTask task = spy(new UploadValidationTask(ctx));
+        task.setHandlerList(handlerList);
+        // return a null record
+        healthDataService = mock(HealthDataService.class);
+        when(healthDataService.getRecordById(any())).thenReturn(null);
+        task.setHealthDataService(healthDataService);
+        task.run();
+        verify(task).logClassNotFoundException(any());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java]: Would it be more useful to have a verify that logDuplicateUploadRecords *isn't* called?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 23]
```diff
@@ -1,30 +1,40 @@
...
 public class DynamoHealthDataDaoDdbTest {
     private static final long CREATED_ON = 1424136378727L;
-    private static final String HEALTH_CODE = "health-code";
-    private static final String RECORD_ID = "record-id";
+    private static final String HEALTH_CODE = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
+    private static final String RECORD_ID = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
     private static final String DATA_TEXT = "{\"data\":\"dummy value\"}";
     private static final String METADATA_TEXT = "{\"metadata\":\"dummy meta value\"}";
-    private static final String SCHEMA_ID = "schema-id";
+    private static final String SCHEMA_ID = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: Randomizing the schema ID is probably not necessary, since nothing here uses it as a key.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 24]
```diff
@@ -86,4 +99,62 @@ public void ddbRoundtrip() throws Exception {
...
+        testRecord.setUserExternalId(USER_EXTERNAL_ID);
+        testRecord.setUserDataGroups(TestConstants.USER_DATA_GROUPS);
+        testRecord.setData(BridgeObjectMapper.get().readTree(DATA_TEXT));
+        testRecord.setMetadata(BridgeObjectMapper.get().readTree(METADATA_TEXT));
+        testRecord.setStudyId(TestConstants.TEST_STUDY_IDENTIFIER);
+
+        mapper.save(testRecord);
+
+        // query that record
+        List<HealthDataRecord> retList = dao.getRecordsByHealthCodeCreatedOnSchemaId(HEALTH_CODE, CREATED_ON, SCHEMA_ID);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: getRecordsByHealthCodeCreatedOnSchemaId() should be moved inside the try-finally. This way, if it throws, the test still cleans up after itself.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 25]
```diff
@@ -86,4 +99,62 @@ public void ddbRoundtrip() throws Exception {
...
+        testRecord.setUploadedOn(UPLOADED_ON);
+        testRecord.setUploadId(UPLOAD_ID);
+        testRecord.setUserSharingScope(ParticipantOption.SharingScope.SPONSORS_AND_PARTNERS);
+        testRecord.setUserExternalId(USER_EXTERNAL_ID);
+        testRecord.setUserDataGroups(TestConstants.USER_DATA_GROUPS);
+        testRecord.setData(BridgeObjectMapper.get().readTree(DATA_TEXT));
+        testRecord.setMetadata(BridgeObjectMapper.get().readTree(METADATA_TEXT));
+        testRecord.setStudyId(TestConstants.TEST_STUDY_IDENTIFIER);
+
+        mapper.save(testRecord);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java:80]: It might be worthwhile to move record creation to an @Before and record deleting to an @After. This removes duplicate code, and if the test fails and then the record deletion also fails, we get a more useful error message.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 26]
```diff
@@ -1,89 +1,193 @@
...
     private static final LocalDate UPLOAD_DATE = LocalDate.parse("2016-05-06");
     private static final String UPLOAD_ID = "upload-id";
     private static final long UPLOADED_ON = 1462575525894L;
     private static final String USER_EXTERNAL_ID = "external-id";
 
+    private static final String HEALTH_CODE_2 = "test-healthcode-2";
+    private static final String SCHEMA_ID_2 = "test-schema-id-2";
+
+    private String HEALTH_CODE;
+    private String RECORD_ID;
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: Style note: Non-static members should be camel-case: healthCode, recordId

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 27]
```diff
@@ -1,89 +1,193 @@
...
+
+//    private String SCHEMA_ID;
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: We can remove this unnecessary comment.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java [hunk 28]
```diff
@@ -1,89 +1,193 @@
...
     private DynamoDBMapper mapper;
 
+    @Autowired
+    private DynamoHealthDataDao dao;
+
+    @Before
+    public void setup() throws IOException {
+        HEALTH_CODE = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
+        RECORD_ID = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
+//        SCHEMA_ID = TestUtils.randomName(DynamoHealthDataDaoDdbTest.class);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDaoDdbTest.java]: Also, this comment.

File: app/org/sagebionetworks/bridge/upload/UploadValidationTask.java [hunk 29]
```diff
@@ -111,6 +124,65 @@ public void run() {
...
+        logger.info("Origin Record: " + "record id: " + originRecord.getId());
+
+        for (HealthDataRecord dupeRecord: dupeRecords) {
+            if (!dupeRecord.getId().equals(originRecord.getId())) {
+                logger.info("Duplicate HealthDataRecord: " + "record id: " + dupeRecord.getId()
+                        + ", createdOn: " + dupeRecord.getCreatedOn().toString()
+                        + ", schemaId: " + dupeRecord.getSchemaId()
+                        + ", studyId: " + dupeRecord.getStudyId() + ", uploadDate: " + dupeRecord.getUploadDate().toString()
+                        + ", uploadId: " + dupeRecord.getUploadId());
+            }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadValidationTask.java:183]: We want to log the original record ID and the dupe record ID in the same log line. Otherwise, it'll be difficult to match up what record is a duplicate of what record.

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 30]
```diff
@@ -3,37 +3,191 @@
...
+
+    private final List<UploadValidationHandler> handlerList = ImmutableList.of(
+            new MessageHandler("foo was here"), new MessageHandler("bar was here"),
+            new MessageHandler("kilroy was here"), new RecordIdHandler(RECORD_ID));
+
+
+    private HealthDataService healthDataService;
+
+    public UploadValidationTaskTest() throws IOException {
+    }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java]: Test classes shouldn't have constructors.

Also, we need a unit test for when context.getRecordId() is null.

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 31]
```diff
@@ -3,37 +3,191 @@
...
+//                .withUploadId(UPLOAD_ID)
+//                .withUserSharingScope(ParticipantOption.SharingScope.SPONSORS_AND_PARTNERS)
+//                .withUserExternalId(USER_EXTERNAL_ID)
+//                .withUserDataGroups(TestConstants.USER_DATA_GROUPS)
+//                .withData(BridgeObjectMapper.get().readTree(DATA_TEXT))
+//                .withMetadata(BridgeObjectMapper.get().readTree(METADATA_TEXT))
+//                .build();
+//
+//        TEST_RECORD_DUPE_LIST_NORMAL = Arrays.asList(TEST_RECORD, TEST_RECORD_DUPE);
+//        TEST_RECORD_DUPE_LIST_MULTI = Arrays.asList(TEST_RECORD, TEST_RECORD_DUPE, TEST_RECORD_DUPE_2);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java]: Why are these commented out?

File: test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java [hunk 32]
```diff
@@ -150,6 +306,131 @@ public void writeValidationStatusException() {
...
+
+        // set up validation task
+        UploadValidationTask task = spy(new UploadValidationTask(ctx));
+        task.setHandlerList(handlerList);
+        // return a null record
+        healthDataService = mock(HealthDataService.class);
+        when(healthDataService.getRecordById(any())).thenReturn(null);
+        task.setHealthDataService(healthDataService);
+        task.run();
+//        verify(task).logClassNotFoundException(any());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/upload/UploadValidationTaskTest.java]: Remove commented out code.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

**Review by DwayneJengSage — APPROVED** [review]
